services/src/main/java/org/keycloak/services/resources/account/AccountCredentialResource.java

```diff
@@ -228,29 +226,23 @@ public CredentialTypeMetadata getMetadata() {
     // Going through all authentication flows and their authentication executions to see if there is any authenticator of the corresponding
     // credential type.
     private Set<String> getEnabledCredentialTypes(List<CredentialProvider> credentialProviders) {
-        Set<String> enabledCredentialTypes = new HashSet<>();
-
-        for (AuthenticationFlowModel flow : realm.getAuthenticationFlows()) {
-            // Ignore DISABLED executions and flows
-            if (isFlowEffectivelyDisabled(flow)) continue;
-
-            for (AuthenticationExecutionModel execution : realm.getAuthenticationExecutions(flow.getId())) {
-                if (execution.getAuthenticator() != null && DISABLED != execution.getRequirement()) {
-                    AuthenticatorFactory authenticatorFactory = (AuthenticatorFactory) session.getKeycloakSessionFactory().getProviderFactory(Authenticator.class, execution.getAuthenticator());
-                    if (authenticatorFactory != null && authenticatorFactory.getReferenceCategory() != null) {
-                        enabledCredentialTypes.add(authenticatorFactory.getReferenceCategory());
-                    }
-                }
-            }
-        }
+        Stream<String> enabledCredentialTypes = realm.getAuthenticationFlowsStream()
+                .filter(((Predicate<AuthenticationFlowModel>) this::isFlowEffectivelyDisabled).negate())
+                .flatMap(flow ->
```

Comment:
We have here a stream of streams and we need to return just flat set of strings.

Reply:
Yeah, but that could have been handled with map only, no? I was mainly curious why flatMap (instead of map) is used since I don't see the need to ensure the underlying stream gets closed, but I might be missing something.

Reply:
Sorry @martin-kanis, I misinterpreted the code. You are right, we have a stream of streams, so we need to flatten them into a single one. Nothing to change here.